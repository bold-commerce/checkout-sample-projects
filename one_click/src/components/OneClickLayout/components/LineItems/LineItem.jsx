import React from 'react';
import PropTypes from 'prop-types';
import Product from '../Product/Product';
import { Price } from '@boldcommerce/stacks-ui'

const LineItem = ({
  title,
  image,
  quantity,
  totalPrice,
  variants,
}) => (
  <div className="SummaryBlock CartItem" >
    <Product
      title={title}
      image={image}
      variants={variants}
    />
    <div className="CartItem__QuantityPriceWrapper">
      <div className="CartItem__ProductQuantityWrapper">
        <div className="ProductQuantity" aria-label="product quantity">
          {quantity}
        </div>
      </div>
      <div className="CartItem__ProductPrice">
        <Price amount={totalPrice} />
      </div>
    </div>
  </div>
);

LineItem.propTypes = {
  title: PropTypes.string,
  image: PropTypes.string,
  quantity: PropTypes.number,
  lineItemKey: PropTypes.string,
  variants: PropTypes.array,
};

export default LineItem;
