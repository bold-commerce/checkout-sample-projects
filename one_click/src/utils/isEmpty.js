const isEmpty = (obj) => {
  return JSON.stringify(obj) === JSON.stringify({});
};

export default isEmpty;