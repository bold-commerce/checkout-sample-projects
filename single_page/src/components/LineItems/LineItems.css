.LineItemQuantity__Decrement button.stx-button,
.LineItemQuantity__Increment button.stx-button {
  background-color: #e7e7e7;
  color: #848484;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.LineItem {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 20px 0;
  border-top: 1px solid rgba(0,0,0,0.1);
}

.LineItem:first-child {
  border-top: none;
  padding-top: 0;
}

.LineItem__ImageDescriptionWrapper {
  flex: 1;
  display: flex;
  justify-content: flex-start;
  align-items: center;
}

.LineItem__QuantityPriceWrapper {
  flex: 1;
  display: flex;
  justify-content: space-between;
}

.LineItem__ProductImageWrapper {
  padding: 5px;
  border: 1px solid #6b6b6b;
  position: relative;
}

.LineItem__ProductImageWrapper .stx-image {
  width: 90px;
  height: 90px;
}

.LineItem__ProductImageWrapper .stx-image--content {
  display: flex;
  justify-content: center;
  align-items: center;
}

.LineItem .LineItem__LineItemQuantityWrapper,
.LineItem .LineItem__ProductPrice {
  display: flex;
  justify-content: center;
  align-items: center;
}

.LineItem .LineItem__ProductDetails {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-start;
  padding: 0 20px;
}

.LineItem .LineItem__ProductDetails .stx-details__description {
  font-size: 16px;
  line-height: 19px;
  color: rgba(0, 0, 0, 0.4);
}

.LineItem .LineItem__ProductDetails .stx-details__title {
  font-size: 20px;
  line-height: 24px;
}

.LineItem .LineItem__ProductDetails a {
  padding: 0;
  margin: 0;
  text-transform: capitalize;
}

.LineItem .LineItem__ProductDetails .stx-button {
  padding: 0;
  margin: 7px 0 0 0;
  text-transform: capitalize;
  text-decoration: underline;
}

.LineItem__ProductPrice {
  font-size: 18px;
  line-height: 21px;
  color: #000000;
  min-width: 65px;
}

.LineItemQuantity__Decrement,
.LineItemQuantity__Increment {
  width: 24px;
  height: 25px;
}

.LineItemQuantity__Decrement .stx-button__text,
.LineItemQuantity__Increment .stx-button__text {
  overflow: visible;
  line-height: 0;
}

.LineItemQuantity {
  width: 40px;
  height: 25px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  font-weight: 500;
  font-size: 14px;
  line-height: 16px;
  display: flex;
  align-items: stretch;
}

.LineItemQuantity input {
  padding: 0.275rem;
  text-align: center;
}

.LineItemQuantity .stx-input {
  height: 100%;
  box-shadow: none;
}

.LineItemQuantity__Decrement button.stx-button,
.LineItemQuantity__Increment button.stx-button,
.LineItemQuantity button.stx-button {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.LineItemQuantity__Decrement,
.LineItemQuantity__Increment,
.LineItemQuantity {
  display: flex;
  justify-content: center;
  align-items: center;
}

.LineItemQuantity--read-only {
  border-radius: 12px;
  background: #a9a9a9;
  color: #fff;
  height: 24px;
  width: 24px;
}

.LineItem__QuantityAdjustmentWrapper {
  flex: 1;
  display: flex;
  align-items: center;
  gap: 10px;
}

.LineItem__QuantityAdjustmentWrapper .LineItemQuantity__Decrement, .LineItem__QuantityAdjustmentWrapper .LineItemQuantity__Increment {
  display: none;
}

.LineItem__SoldOut {
  background: #ffebf0;
  padding: 5px 10px;
  text-transform: uppercase;
  font-size: 12px;
  font-weight: bold;
  color: #b4002b;
  border: 1px solid #b4002b;
}

@media screen and (max-width: 599px) {
  .LineItem__ImageDescriptionWrapper {
    flex: 1.5;
  }

  .LineItem__QuantityPriceWrapper {
    flex-direction: column;
    align-items: flex-end;
    justify-content: space-around;
  }

  .LineItem__ProductImageWrapper {
    padding: 3px;
  }

  .LineItem__ProductImageWrapper .stx-image {
    width: 54px;
    height: 54px;
  }

  .LineItem .LineItem__ProductDetails {
    padding: 0 10px;
  }

  .LineItem .LineItem__ProductDetails .stx-details__description {
    font-size: 12px;
    line-height: 14px;
  }

  .LineItem .LineItem__ProductDetails .stx-details__title {
    font-size: 14px;
    line-height: 17px;
  }

  .LineItem .LineItem__ProductDetails a {
    font-size: 12px;
    line-height: 16px;
  }

  .LineItem__ProductPrice .stx-price__amount {
    margin: unset;
    font-style: normal;
    font-weight: normal;
    font-size: 14px;
    line-height: 17px;
  }
}
