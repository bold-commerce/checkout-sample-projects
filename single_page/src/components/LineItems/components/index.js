export { default as LineItem } from './LineItem';
export { default as Product } from './Product';
