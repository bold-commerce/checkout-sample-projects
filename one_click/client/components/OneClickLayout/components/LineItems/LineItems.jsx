/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import LineItem from './LineItem';
import { useLineItems } from '@boldcommerce/checkout-react-components';
import { useVariants } from '../../../../hooks';

export const LineItems = ({
  readOnly,
  lineItems,
  updateLineItemQuantity,
  removeLineItem,
}) => {
  const handleVariants = useVariants();
  const lineItemList = lineItems.map((item) => (
    <LineItem
      title={item.product_data.product_title}
      description={handleVariants(item.product_data.title)}
      image={item.product_data.image_url}
      quantity={item.product_data.quantity}
      price={item.product_data.price}
      totalPrice={item.product_data.total_price}
      lineItemKey={item.product_data.line_item_key}
      onQuantityChange={(lineItemKey, value) => updateLineItemQuantity(lineItemKey, value)}
      onRemove={(lineItemKey) => removeLineItem(lineItemKey)}
      readOnly={readOnly}
      key={item.product_data.line_item_key}
    />
  ));

  return lineItemList;
};

LineItems.propTypes = {
  readOnly: PropTypes.bool,
  lineItems: PropTypes.array,
  updateLineItemQuantity: PropTypes.func,
  removeLineItem: PropTypes.func,
};

const MemoizedLineItems = React.memo(LineItems);

const LineItemsContainer = () => {
  const { data: lineItems } = useLineItems();

  return (
    <MemoizedLineItems
      lineItems={lineItems}
    />
  );
};

LineItemsContainer.propTypes = {
  readOnly: PropTypes.bool,
};

export default LineItemsContainer;
