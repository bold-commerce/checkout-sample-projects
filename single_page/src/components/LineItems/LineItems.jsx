import React, { memo, useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';
import { useLineItems } from '@boldcommerce/checkout-react-components';
import { LineItem } from './components';
import { useErrorLogging } from '../../hooks';
import { debounce } from '../../helpers/debounce';
import './LineItems.css';

const LineItems = ({ readOnly = false }) => {
  const { data, updateLineItemQuantity, removeLineItem } = useLineItems();

  return (
    <MemoizedLineItems
      lineItems={data}
      onUpdate={updateLineItemQuantity}
      onRemove={removeLineItem}
      readOnly={readOnly}
    />
  );
};

const MemoizedLineItems = memo(({
  lineItems,
  onUpdate,
  onRemove,
  readOnly,
}) => {
  const logError = useErrorLogging();

  const handleUpdate = useCallback(async (lineItemKey, value) => {
    try {
      await onUpdate(lineItemKey, value);
    } catch(e) {
      logError('line_items', e);
    }
  }, []);

  const debouncedUpdate = useMemo(() => debounce(handleUpdate, 300), [handleUpdate]);

  const handleRemove = useCallback(async (lineItemKey) => {
    try {
      await onRemove(lineItemKey);
    } catch(e) {
      logError('line_items', e);
    }
  });

  const lineItemList = lineItems.map((item) => (
    <LineItem
      title={item.product_data.title}
      image={item.product_data.image_url}
      quantity={item.product_data.quantity}
      price={item.product_data.price}
      totalPrice={item.product_data.total_price}
      lineItemKey={item.product_data.line_item_key}
      onChange={(lineItemKey, value) => debouncedUpdate(lineItemKey, value)}
      onRemove={(lineItemKey) => handleRemove(lineItemKey)}
      readOnly={readOnly}
      key={item.product_data.line_item_key}
    />
  ));

  return (
    <div className="LineItemList">
      {lineItemList}
    </div>
  );
});

LineItems.propTypes = {
  readOnly: PropTypes.bool,
};

export default LineItems;
