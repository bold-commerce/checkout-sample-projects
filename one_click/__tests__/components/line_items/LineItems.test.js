import React from 'react';
import { render } from '@testing-library/react';
import { LineItems } from '../../../src/components/OneClickLayout/components/LineItems/LineItems'
import { exampleLineItems } from '../../utils/lineItemHelpers';

describe('LineItems', () => {
  test('renders LineItems component with editable quantity', () => {
    const { asFragment } = render(
      <LineItems
        readOnly={false}
        lineItems={exampleLineItems}
        updateLineItemQuantity={() => {}}
        removeLineItem={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders read only LineItems component', () => {
    const { asFragment } = render(
      <LineItems
        readOnly
        lineItems={exampleLineItems}
        updateLineItemQuantity={() => {}}
        removeLineItem={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
