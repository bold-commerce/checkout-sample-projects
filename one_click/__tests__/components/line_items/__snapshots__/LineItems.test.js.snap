// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineItems renders LineItems component with editable quantity 1`] = `
<DocumentFragment>
  <div
    class="SummaryBlock CartItem"
  >
    <div
      class="CartItem__ImageDescriptionWrapper"
    >
      <div
        class="CartItem__ProductImageWrapper"
      >
        <div
          class="stx-image"
        >
          <div
            class="stx-image--content"
          >
            <div
              class="stx-image--holder"
            >
              <img
                alt="Test Product A"
                class="stx-image--image"
                src="https://via.placeholder.com/150"
                title="Test Product A"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="CartItem__ProductDetails"
      >
        <div
          class="stx-details stx-details--left"
        >
          <div
            class="stx-details__title"
          >
            Test Product A
          </div>
          <div
            class="stx-details__description"
          />
        </div>
      </div>
    </div>
    <div
      class="CartItem__QuantityPriceWrapper"
    >
      <div
        class="CartItem__ProductQuantityWrapper"
      >
        <div
          class="ProductQuantity"
        >
          2
        </div>
      </div>
      <div
        class="CartItem__ProductPrice"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$59.98
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="SummaryBlock CartItem"
  >
    <div
      class="CartItem__ImageDescriptionWrapper"
    >
      <div
        class="CartItem__ProductImageWrapper"
      >
        <div
          class="stx-image"
        >
          <div
            class="stx-image--content"
          >
            <div
              class="stx-image--holder"
            >
              <img
                alt="Test Product B"
                class="stx-image--image"
                src="https://via.placeholder.com/150"
                title="Test Product B"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="CartItem__ProductDetails"
      >
        <div
          class="stx-details stx-details--left"
        >
          <div
            class="stx-details__title"
          >
            Test Product B
          </div>
          <div
            class="stx-details__description"
          />
        </div>
      </div>
    </div>
    <div
      class="CartItem__QuantityPriceWrapper"
    >
      <div
        class="CartItem__ProductQuantityWrapper"
      >
        <div
          class="ProductQuantity"
        >
          4
        </div>
      </div>
      <div
        class="CartItem__ProductPrice"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$79.96
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`LineItems renders read only LineItems component 1`] = `
<DocumentFragment>
  <div
    class="SummaryBlock CartItem"
  >
    <div
      class="CartItem__ImageDescriptionWrapper"
    >
      <div
        class="CartItem__ProductImageWrapper"
      >
        <div
          class="stx-image"
        >
          <div
            class="stx-image--content"
          >
            <div
              class="stx-image--holder"
            >
              <img
                alt="Test Product A"
                class="stx-image--image"
                src="https://via.placeholder.com/150"
                title="Test Product A"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="CartItem__ProductDetails"
      >
        <div
          class="stx-details stx-details--left"
        >
          <div
            class="stx-details__title"
          >
            Test Product A
          </div>
          <div
            class="stx-details__description"
          />
        </div>
      </div>
    </div>
    <div
      class="CartItem__QuantityPriceWrapper"
    >
      <div
        class="CartItem__ProductQuantityWrapper"
      >
        <div
          class="ProductQuantity"
        >
          2
        </div>
      </div>
      <div
        class="CartItem__ProductPrice"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$59.98
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="SummaryBlock CartItem"
  >
    <div
      class="CartItem__ImageDescriptionWrapper"
    >
      <div
        class="CartItem__ProductImageWrapper"
      >
        <div
          class="stx-image"
        >
          <div
            class="stx-image--content"
          >
            <div
              class="stx-image--holder"
            >
              <img
                alt="Test Product B"
                class="stx-image--image"
                src="https://via.placeholder.com/150"
                title="Test Product B"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="CartItem__ProductDetails"
      >
        <div
          class="stx-details stx-details--left"
        >
          <div
            class="stx-details__title"
          >
            Test Product B
          </div>
          <div
            class="stx-details__description"
          />
        </div>
      </div>
    </div>
    <div
      class="CartItem__QuantityPriceWrapper"
    >
      <div
        class="CartItem__ProductQuantityWrapper"
      >
        <div
          class="ProductQuantity"
        >
          4
        </div>
      </div>
      <div
        class="CartItem__ProductPrice"
      >
        <div
          class="stx-price stx-price--alight-center"
        >
          <p
            aria-label="Product price"
            class="stx-price__amount"
          >
            CA$79.96
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
